xarray/backends/api.py

```diff
@@ -542,7 +527,99 @@ def maybe_decode_store(store, lock=False):
+            ds2 = ds.chunk(chunks, name_prefix=name_prefix, token=token)
+            ds2._file_obj = ds._file_obj
+
+        else:  # file is zarr!
+
+            # adapted from Dataset.Chunk() and taken from open_zarr
+            if not isinstance(chunks, (int, dict)):
+                if chunks != "auto":
+                    raise ValueError(
+                        "chunks must be an int, dict, 'auto', or None. "
+                        "Instead found %s. " % chunks
+                    )
+            if isinstance(chunks, int):
+                chunks = dict.fromkeys(ds.dims, chunks)
+
+            if isinstance(chunks, tuple) and len(chunks) == len(ds.dims):
+                chunks = dict(zip(ds.dims, chunks))
+
+            def get_chunk(name, var, chunks):
+                chunk_spec = dict(zip(var.dims, var.encoding.get("chunks")))
+
+                # Coordinate labels aren't chunked
+                if var.ndim == 1 and var.dims[0] == name:
+                    return chunk_spec
+
+                if chunks == "auto":
+                    return chunk_spec
+
+                for dim in var.dims:
+                    if dim in chunks:
+                        spec = chunks[dim]
+                        if isinstance(spec, int):
+                            spec = (spec,)
+                        if isinstance(spec, (tuple, list)) and chunk_spec[dim]:
+                            if any(s % chunk_spec[dim] for s in spec):
+                                warnings.warn(
+                                    "Specified Dask chunks %r would "
+                                    "separate Zarr chunk shape %r for "
+                                    "dimension %r. This significantly "
+                                    "degrades performance. Consider "
+                                    "rechunking after loading instead."
+                                    % (chunks[dim], chunk_spec[dim], dim),
+                                    stacklevel=2,
+                                )
+                        chunk_spec[dim] = chunks[dim]
+                return chunk_spec
+
+            def maybe_chunk(name, var, chunks):
+                chunk_spec = get_chunk(name, var, chunks)
+
+                if (var.ndim > 0) and (chunk_spec is not None):
+                    # does this cause any data to be read?
+                    token2 = tokenize(name, var._data)
+                    name2 = "zarr-%s" % token2
+                    var = var.chunk(chunk_spec, name=name2, lock=None)
+                    if overwrite_encoded_chunks and var.chunks is not None:
+                        var.encoding["chunks"] = tuple(x[0] for x in var.chunk)
+                    return var
+                else:
+                    return var
+
+            variables = {k: maybe_chunk(k, v, chunks) for k, v in ds.variables.items()}
+            ds2 = ds._replace_vars_and_dims(variables)
+        return ds2
+    else:
+        ds2 = ds
+
+    return ds2
```

Comment:
since we return in the block above, I think it would be okay to also return in line 620.

Reply:
This part I am not so sure I completely understand. ``ds2`` gets created if ``chunk`` is not ``None``. If ``chunk is None`` then ``ds`` is passed, so we need ``ds2=ds``.

This brings to a question: Can we just have
```python
ds = ds.chunk(chunks, ...)
```
instead of 
```python 
ds2 = ds.chunk(chunks,...)
ds2._file_obj = ds._file_obj
```
that way I could just return ``ds`` on line ``620``